### PR TITLE
ddl: year type should have an unsigned flag

### PR DESCRIPTION
**Pull request by jackysp**
### What problem does this PR solve? 
Fix https://github.com/pingcap/tidb/issues/7537. year type should have an unsigned flag. According to https://dev.mysql.com/doc/refman/8.0/en/numeric-type-overview.html ,
if you specify ZEROFILL for a numeric column, MySQL automatically adds the UNSIGNED attribute to the column.

### What is changed and how it works?
Add unsigned flag back to year type, and add zerofill flag for all types except year type in `show create table`.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

PTAL @zimulala @winkyao

### CODE REVIEW

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -394,6 +394,12 @@ func columnDefToCol(ctx sessionctx.Context, offset int, colDef *ast.ColumnDef, o
 		col.Flag &= ^mysql.BinaryFlag
 		col.Flag |= mysql.ZerofillFlag
 	}
+	// If you specify ZEROFILL for a numeric column, MySQL automatically adds the UNSIGNED attribute to the column.
+	// See https://dev.mysql.com/doc/refman/8.0/en/numeric-type-overview.html for more details.
+	// But some types like bit and year, won't show its unsigned flag in `show create table`.
```
Comment (lysu) [inline ddl/ddl_api.go]: we should not support https://github.com/pingcap/tidb/issues/6688?

File: executor/show_test.go [hunk 2]
```diff
@@ -369,6 +369,16 @@ func (s *testSuite) TestShow(c *C) {
 			"  `x` int(11) DEFAULT NULL,\n"+
 			"  PRIMARY KEY (`y`)\n"+
 			") ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin"))
+
+	// Test show create table with zerofill flag
+	tk.MustExec(`drop table if exists t`)
+	tk.MustExec(`create table t(id int primary key, val tinyint(10) zerofill);`)
```
Comment (winkyao) [inline executor/show_test.go]: Add a year type?

Comment (jackysp) [reply executor/show_test.go]: year with zerofill?

File: ddl/ddl_api.go (hunk 1 above)
Comment (jackysp) [reply ddl/ddl_api.go]: It seems it still could work in some client/driver, see https://github.com/pingcap/tidb/pull/7525

**Comment by jackysp** [comment]
@winkyao in this pr https://github.com/pingcap/tidb/pull/6745 , I've tested the schema in it, but MySQL  does not return an error.
```
mysql> CREATE TABLE `abc` (
    ->   `y` year UNSIGNED NOT NULL,
    ->   `x` int(11) DEFAULT NULL,
    ->   PRIMARY KEY (`y`)
    -> ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin
    -> ;
Query OK, 0 rows affected
```
I think we should fix it in parser.

File: executor/show_test.go (hunk 2 above)
Comment (jackysp) [reply executor/show_test.go]: It is not allowed, according to https://github.com/pingcap/tidb/pull/7542#issuecomment-416974493 @shenli

**Comment by jackysp** [comment]
PTAL @winkyao

**Comment by jackysp** [comment]
Add the parser support for year with field option according to https://github.com/mysql/mysql-server/blob/68816f735f02735e10ea6bea8efc291e4d3cac33/sql/sql_yacc.yy#L6263
PTAL @shenli

File: ddl/db_test.go [hunk 3]
```diff
@@ -2157,24 +2157,29 @@ func (s *testDBSuite) TestRebaseAutoID(c *C) {
...
+	s.tk.MustExec("create table abc(y year, z tinyint(10) zerofill, primary key(y));")
 	is := s.dom.InfoSchema()
 	tbl, err := is.TableByName(model.NewCIStr("test"), model.NewCIStr("abc"))
 	c.Assert(err, IsNil)
-	var yearCol *model.ColumnInfo
+	var yearCol, zCol *model.ColumnInfo
 	for _, col := range tbl.Meta().Columns {
 		if col.Name.String() == "y" {
 			yearCol = col
-			break
```
Comment (zhexuany) [inline ddl/db_test.go]: Will this removal affect test-case's behavior?

File: ddl/ddl_api.go (hunk 1 above)
Comment (zhexuany) [reply ddl/ddl_api.go]: myabe we need use `5.7`

File: ddl/db_test.go (hunk 3 above)
Comment (jackysp) [reply ddl/db_test.go]: No.

**Comment by coocood** [comment]
LGTM

**Pull request merged**